clippy_lints/src/option_map_unit_fn.rs

```diff
@@ -0,0 +1,188 @@
+use utils::paths;
+
+#[derive(Clone)]
+pub struct Pass;
+
+/// **What it does:** Checks for usage of `Option.map(f)` where f is a function
+/// or closure that returns the unit type.
+///
+/// **Why is this bad?** Readability, this can be written more clearly with
+/// an if statement
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// let x : Option<&str> = do_stuff();
+/// x.map(log_err_msg);
+/// x.map(|msg| log_err_msg(format_msg(msg)))
+/// ```
+/// The correct use would be:
+/// ```rust
+/// let x : Option<&str> = do_stuff();
+/// if let Some(msg) = x {
+///     log_err_msg(msg)
+/// }
+/// if let Some(msg) = x {
+///     log_err_msg(format_msg(msg))
+/// }
+/// ```
+declare_clippy_lint! {
+    pub OPTION_MAP_UNIT_FN,
+    complexity,
+    "using `Option.map(f)`, where f is a function or closure that returns ()"
+}
+
+
+impl LintPass for Pass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(OPTION_MAP_UNIT_FN)
+    }
+}
+
+fn is_unit_type(ty: ty::Ty) -> bool {
+    match ty.sty {
+        ty::TyTuple(slice) => slice.is_empty(),
+        ty::TyNever => true,
+        _ => false,
+    }
+}
+
+fn is_unit_function(cx: &LateContext, expr: &hir::Expr) -> bool {
+    let ty = cx.tables.expr_ty(expr);
+
+    if let ty::TyFnDef(id, _) = ty.sty {
+        if let Some(fn_type) = cx.tcx.fn_sig(id).no_late_bound_regions() {
+            return is_unit_type(fn_type.output());
+        }
+    }
+    false
+}
+
+fn is_unit_expression(cx: &LateContext, expr: &hir::Expr) -> bool {
+    is_unit_type(cx.tables.expr_ty(expr))
+}
+
+// The expression inside a closure may or may not have surrounding braces and
+// semicolons, which causes problems when generating a suggestion. Given an
+// expression that evaluates to '()' or '!', recursively remove useless braces
+// and semi-colons until is suitable for including in the suggestion template
+fn reduce_unit_expression<'a>(cx: &LateContext, expr: &'a hir::Expr) -> Option<Span> {
+    if !is_unit_expression(cx, expr) {
+        return None;
+    }
+
+    match expr.node {
+        hir::ExprCall(_, _) |
+        hir::ExprMethodCall(_, _, _) => {
+            // Calls can't be reduced any more
+            Some(expr.span)
+        },
+        hir::ExprBlock(ref block) => {
+            match (&block.stmts[..], block.expr.as_ref()) {
+                (&[], Some(inner_expr)) => {
+                    // Reduce `{ X }` to `X`
+                    reduce_unit_expression(cx, inner_expr)
+                },
+                (&[ref inner_stmt], None) => {
+                    // Reduce `{ X; }` to `X` or `X;`
+                    match inner_stmt.node {
+                        hir::StmtDecl(ref d, _) => Some(d.span),
+                        hir::StmtExpr(ref e, _) => Some(e.span),
+                        hir::StmtSemi(ref e, _) => {
+                            if is_unit_expression(cx, e) {
```

Comment:
Any particular reason we don't just keep the semicolon always?

Reply:
I pushed 85ad541 to remove it and it seems fine. The new stderr output doesn't show any weird changes, so I guess it was just an optical thing.